packages/sdk/identity/src/odis/identifier.ts

```diff
@@ -0,0 +1,202 @@
+import {
+  CombinerEndpoint,
+  KEY_VERSION_HEADER,
+  SignMessageRequest,
+  SignMessageResponseSchema,
+} from '@celo/phone-number-privacy-common'
+import { soliditySha3 } from '@celo/utils/lib/solidity'
+import { createHash } from 'crypto'
+import debugFactory from 'debug'
+import { BlsBlindingClient, WasmBlsBlindingClient } from './bls-blinding-client'
+import {
+  AuthenticationMethod,
+  AuthSigner,
+  EncryptionKeySigner,
+  getOdisPnpRequestAuth,
+  queryOdis,
+  ServiceContext,
+} from './query'
+
+const debug = debugFactory('kit:odis:identifier')
+const sha3 = (v: string) => soliditySha3({ type: 'string', value: v })
+
+const PEPPER_CHAR_LENGTH = 13
+const PEPPER_SEPARATOR = '__'
+
+export enum IdentifierType {
+  PHONE_NUMBER = 0,
+  EMAIL = 1,
+  TWITTER = 2,
+  // feel free to put up a PR to add more types!
+}
+
+export function getIdentifierPrefix(type: IdentifierType) {
+  switch (type) {
+    case IdentifierType.PHONE_NUMBER:
+      return 'tel'
+    case IdentifierType.EMAIL:
+      return 'mailto'
+    case IdentifierType.TWITTER:
+      return 'twit'
+    default:
+      throw new Error('Unsupported Identifier Type')
+  }
+}
+
+export interface IdentifierHashDetails {
+  offchainIdentifier: string
+  identifierHash: string
+  pepper: string
+  unblindedSignature?: string
+}
+
+/**
+ * Retrieve the on-chain identifier for the provided off-chain identifier
+ * Performs blinding, querying, and unblinding
+ */
+export async function getOnchainIdentifier(
+  offchainIdentifier: string,
+  identifierType: string | IdentifierType,
+  account: string,
+  signer: AuthSigner,
+  context: ServiceContext,
+  blindingFactor?: string,
+  clientVersion?: string,
+  blsBlindingClient?: BlsBlindingClient,
+  sessionID?: string
+): Promise<IdentifierHashDetails> {
+  debug('Getting identifier pepper')
+
+  let seed: Buffer | undefined
+  if (blindingFactor) {
+    seed = Buffer.from(blindingFactor)
+  } else if (signer.authenticationMethod === AuthenticationMethod.ENCRYPTION_KEY) {
+    seed = Buffer.from((signer as EncryptionKeySigner).rawKey)
+  }
+
+  // Fallback to using Wasm version if not specified
+
+  if (!blsBlindingClient) {
+    debug('No BLSBlindingClient found, using WasmBlsBlindingClient')
+    blsBlindingClient = new WasmBlsBlindingClient(context.odisPubKey)
+  }
+
+  const base64BlindedMessage = await getBlindedIdentifier(
+    offchainIdentifier,
```

Comment:
One concern we discussed at the offsite around this is that without including prefixes in the actual blinded message passed to ODIS (which is IIUC what we're doing here) is that we will get the same pepper for different types of identifiers (i.e. `twit:username` and `insta:username` would lead to the same pepper for `username`). One way of getting around this would be treating phone numbers as special cases (i.e. no prefix in the pepper generation) but including the prefix for other blinded messages, though I personally would want to get additional eyes on this from the crypto team folks about the security implications of this before moving forward with this. Let me know if I'm misunderstanding something about this though!!

Reply:
I was hesitant to change the blinding because I wanted to make sure we are still interoperable with pre-existing phone number identifiers. I'm worried handling the phone numbers differently might be confusing, especially if someone chooses to forgo the sdk and query ODIS/blind/hash themselves.

But definitely still an important point, thanks for pointing it out! How much of a security risk do we think it is for the same plaintext identifier across different types to have the same pepper? @nategraf

Reply:
Doing it this way basically means that if you query once for `username`, you can derive the on-chain identifier for twitter://username  and instagram://username without querying a second time